### DISCUSSION TITLE
What is the best way to type cast os.date('%Y') return value as integer?

### DISCUSSION THREAD

**Discussion by stefanos82** [edubart/nelua-lang#216] [Q&A] [answered]
I have attempted the following example without success:

```lua
require 'os'

local year = (@int64)(os.date('%Y'))
print(year)
```

![image](https://github.com/edubart/nelua-lang/assets/1777466/e03eaf7b-bacf-4d7a-99fe-0a888a1e4ee8)

What is the ideal way to cast a string to integer?

**Comment by RaycatWhoDat** [comment, marked as answer]
```lua
require 'os'

local year: int64 = tonumber(os.date('%Y'))
print(year)
```
?

Also, you'll probably get a faster response in the Discord.

**Reply by stefanos82** [reply]
I have totally forgot about `tonumber` :facepalm: thank you @RayMPerry :+1:

**Comment by stefanos82** [comment]
@edubart based on the example above by @RayMPerry, if I don't set its type explicitly, why is it `nil`?

```lua
require 'os'

do
  local year = tonumber(os.date('%Y'), 10)
  print(year) -- 2023
  ##[[
  print(string.format("year.type: %s", year.type)) -- year.type: nil
  ]]
end
```

I have tried the same with `tointeger()` and still the same.

I thought this was being applied:

![image](https://github.com/edubart/nelua-lang/assets/1777466/d346cab8-c987-4b9a-963f-600bffa69d68)

Am I doing something wrong?

**Reply by RaycatWhoDat** [reply]
According to `tonumber`'s definition [here](https://github.com/edubart/nelua-lang/blob/e084bf1cc8a5651c9fdd96bc4205a890e5ab1fca/lib/string.nelua#L973), it returns `auto`. An `auto` variable doesn't have a `type` property like others. Dunno why that is; I just know it is.

```lua
local function return_auto(): auto
   return 12
end

local other_var = return_auto()
## print(string.format("other_var.type: %s", other_var.type)) -- nil
```

Workaround: just cast it.

**Reply by stefanos82** [reply]
Something is not right...in the comment in my screenshot above it makes it clear about type deduction.

**Reply by stefanos82** [reply]
@RayMPerry Oh shoot, I forgot about the `after_inference()` that can be used to find the types _after_ the compiler finishes the type inference.

```lua
require 'os'

do
  local year = tointeger(os.date('%Y'))
  print(year)
  ##[[
  after_inference(function()
    print(string.format("year.type: %s", year.type))
end)]]
end
```

**Reply by RaycatWhoDat** [reply]
Huh. TIL.

**Reply by stefanos82** [reply]
Yeah, I totally forgot an old example Eduardo shared with me about various ways to get variable types in 5 different ways:

```lua
require 'io'

local a: auto = 1
local b = 1
local c: #[a.type]# = 1

-- at compile time
##[[
print(string.format("At compile-time\n  a: %6s, b: %6s, c: %6s", a.type, b.type, c.type))
]]
-- the above outputs: a: int64, b: nil, c: int64

-- at compile time after type inference is finished
##[[
after_inference(function()
  print(string.format("At compile-time after type inference\n" ..
      "  a: %6s, b: %6s, c: %6s", a.type, b.type, c.type))
end)
]]
-- the above outputs: a: int64, b: int64, c: int64

-- at runtime using preprocessor
io.printf("At runtime using preprocessor\n" .. 
  "  a: %6s, b: %6s, c: %6s\n",
  #[tostring(a.type)]#,
  #[tostring(b.type)]#,
  #[tostring(c.type)]#)
-- the above outputs: a: int64, b: nil, c: int64

-- at runtime using the traits module
require 'traits'
io.printf("At runtime using traits module\n" .. 
  "  a: %6s, b: %6s, c: %6s\n",
  traits.typeinfoof(a).name,
  traits.typeinfoof(b).name,
  traits.typeinfoof(c).name)
-- the above outputs: a: int64, b: int64, c: int64

-- at runtime using a customized polymorphic function
local function typenameof(x: auto): string
  return #[tostring(x.type)]#
end
io.printf("At runtime using polymorphic function\n" .. 
  "  a: %6s, b: %6s, c: %6s\n",
  typenameof(a), typenameof(b), typenameof(c))
-- the above outputs: a: int64, b: int64, c: int64
```
